### PR TITLE
tests: run codegen only when necessary, use correct $GOPATH

### PR DESCRIPTION
**Pull request by stevekuznetsov**
Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

/test pull-test-infra-verify-codegen
/area prow
/kind bug
/cc @BenTheElder @fejta 
/assign @cjwagner

### CODE REVIEW

File: hack/update-codegen.sh [hunk 1]
```diff
@@ -17,6 +17,7 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
+GOPATH="${GOPATH:-"${HOME}/go"}"
```
Comment (BenTheElder) [inline hack/update-codegen.sh]: `go env GOPATH` is probably a better idea if we're just using the `go` in `$PATH` anyhow.

**Comment by stevekuznetsov** [comment]
Hmm I explicitly asked for the test but it's still skipped?

/test pull-test-infra-verify-codegen

**Comment by stevekuznetsov** [comment]
@BenTheElder looks like something is providing a `$GOPATH` of `/go` to the script, which is wrong. What mechanism is that?

**Comment by stevekuznetsov** [comment]
```
W0726 20:11:18.323] + echo 'Pre-existing $GOPATH: /go'
W0726 20:11:18.333] + echo 'Go env: /go'
W0726 20:11:18.335] + echo 'Current working directory: /workspace/k8s.io/test-infra/hack'
```

File: config/jobs/kubernetes/test-infra/test-infra-presubmits.yaml [hunk 2]
```diff
@@ -205,7 +205,7 @@ presubmits:
     - master
     rerun_command: "/test pull-test-infra-verify-codegen"
     trigger: "/test pull-test-infra-verify-codegen"
-    run_if_changed: '^(Gopkg\.|^vendor/).*$'
+    run_if_changed: 'prow/(apis|client)/.*$'
```
Comment (fejta) [inline config/jobs/kubernetes/test-infra/test-infra-presubmits.yaml]: Also vendor

File: hack/update-codegen.sh [hunk 3]
```diff
@@ -16,7 +16,13 @@
 set -o errexit
 set -o nounset
 set -o pipefail
+set -o xtrace
 
+echo "Pre-existing \$GOPATH: ${GOPATH}"
```
Comment (fejta) [inline hack/update-codegen.sh]: You need ${GOPATH;-} else this will blow up because of `nounset`, right?

**Comment by ixdy** [comment]
yeah, it seems like the job is not configured to clone the repository into the GOPATH.

I think we may do this correctly for k/k, though I'm not entirely sure.

**Comment by ixdy** [comment]
actually, I'm not sure we even handle k/k properly. maybe we get away with it due to docker-in-docker?

**Comment by ixdy** [comment]
ah, I think the answer is that you need to pass `--root=/go/src` to bootstrap.

File: hack/update-codegen.sh (hunk 3 above)
Comment (stevekuznetsov) [reply hack/update-codegen.sh]: This is debug logging to figure out why my script is getting `$GOPATH=/go`

**Comment by stevekuznetsov** [comment]
ewwww

**Comment by fejta** [comment]
Why are you using bootstrap?!

**Comment by stevekuznetsov** [comment]
Yeah I guess I could just decorate, I started out trying to do some `bazel` stuff

**Comment by fejta** [comment]
bazel is no reason not to use the decorators. Just requires an image with bazel in it. Note that the rules_k8s job does not use bootstrap and (obviously) uses bazel: https://github.com/kubernetes/test-infra/blob/master/config/jobs/bazelbuild/rules_k8s/rules_k8s_config.yaml

Image def: https://github.com/bazelbuild/rules_k8s/tree/master/images/gcloud-bazel

**Comment by fejta** [comment]
But the decorators are SUPER NICE!! You should refactor the job to use them and give kudos to whoever creates them :)

**Comment by fejta** [comment]
Worst case scenario: job fails 100% just like it does today :)

**Comment by stevekuznetsov** [comment]
OK, changed to decorators but kept optional, lets see how much more broken this gets
/hold cancel

**Pull request merged**